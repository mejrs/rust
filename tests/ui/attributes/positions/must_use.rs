//@aux-build:dummy.rs

#![forbid(unstable_features)]
#![warn(unused_attributes)]

#![must_use] //~ WARN

#[must_use] //~ WARN
extern crate dummy;

extern crate core;

#[must_use] //~ WARN
pub mod empty_crate;

#[must_use] //~ WARN
pub mod module {

    #[must_use] //~ WARN
    pub static GLOBAL: u32 = 42;

    #[must_use] //~ WARN
    pub static mut GLOBAL_MUT: u32 = 42;

    #[must_use] //~ WARN
    pub const CONST: u32 = 42;

    #[must_use] //~ WARN
    use ::core::fmt::Debug;

    #[must_use] //~ WARN
    trait TraitAlias = Debug; //~ ERROR trait aliases are experimental

    #[must_use] //~ WARN
    pub type TypeAlias = u32;

    #[must_use] // OK
    pub struct Struct<
        #[must_use] 'lt, //~ WARN
        #[must_use] T> //~ WARN
     {
        #[must_use] //~ WARN
        pub struct_field: u32,
        #[must_use] //~ WARN
        pub generic_field: &'lt T,
    }

    #[must_use] // OK
    pub struct TupleStruct(#[must_use] pub u32); //~ WARN

    #[must_use] // OK
    pub enum Enum {
        #[must_use] //~ WARN
        FieldVariant {
            #[must_use] //~ WARN
            field: u32,
        },
        #[must_use] //~ WARN
        TupleVariant(#[must_use] u32), //~ WARN
    }
    #[must_use] // OK
    pub union Union {
        #[must_use] //~ WARN
        pub field: u32,
    }

    #[must_use] //~ WARN
    impl<'lt, T> Struct<'lt, T> {

        #[must_use] // OK
        pub fn method(#[must_use] &mut self) { //~ ERROR allow, cfg, cfg_attr, deny, expect, forbid, and warn are the only allowed built-in attributes in function parameters
            //~^ WARN
            #[must_use] //~ ERROR
            //~^ WARN
            self.struct_field = 73;
        }

        #[must_use] // OK
        pub fn static_method(#[must_use] _arg: u32) {} //~ ERROR allow, cfg, cfg_attr, deny, expect, forbid, and warn are the only allowed built-in attributes in function parameters
        //~^ WARN
    }

    #[must_use] //~ WARN
    impl<
        #[must_use] 'lt, //~ WARN
        #[must_use] T //~ WARN
    > Drop for Struct<'lt, T> {
        #[must_use] //~ WARN
        fn drop(&mut self) {
            // ..
        }
    }

    #[must_use] // OK
    pub fn function<
        #[must_use] const N: usize //~ WARN
    >(arg: u32) -> u32 {
        #![must_use] //~ WARN
        //~^ WARN this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!

        arg
    }

    #[must_use] // OK
    pub trait Trait {
        #[must_use] // OK
        fn trait_method() {}
    }

    fn x() {
        #[must_use] //~ WARN
        const {
            // ..
        }
    }
}

pub mod inner_module {
    #![must_use] //~ WARN

    impl<'lt, T> super::module::Struct<'lt, T> {
        #![must_use] //~ WARN

        pub fn inner_method(&mut self) {
            #![must_use] // OK
        }
    }

    fn x() {
        const {
            #![must_use] //~ WARN
        }
    }
}

#[must_use] // OK
fn main() {
    let _closure = #[must_use] //~ ERROR attributes on expressions are experimental
    //~^ WARN
    | #[must_use] _arg: u32| {}; //~ ERROR
    //~^ WARN
    let _move_closure = #[must_use] //~ ERROR attributes on expressions are experimental
    //~^ WARN
    move | #[must_use] _arg: u32| {}; //~ ERROR
    //~^ WARN
    #[must_use] //~ WARN
    {
        #![must_use] //~ WARN
        //~^ WARN unused attribute
        //~^^ WARN this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!

        #[must_use] //~ WARN
        let variable = 42_u32;

        let _array = #[must_use] //~ ERROR attributes on expressions are experimental
        //~^ WARN
        [
            #[must_use] //~ WARN
            1,
            2,
        ];
        let _tuple = #[must_use] //~ ERROR attributes on expressions are experimental
        //~^ WARN
        (
            #[must_use] //~ WARN
            1,
            2,
        );
        let _tuple_struct = module::TupleStruct(
            #[must_use] //~ WARN
            2,
        );
        let _struct = module::Struct {
            #[must_use] //~ WARN
            struct_field: 42,
            generic_field: &13,
        };

        let _union = module::Union {
            #[must_use] //~ WARN
            field: 42,
        };

        let _fn_call = module::function::<7>(
            #[must_use] //~ WARN
            42,
        );

        #[must_use] //~ WARN
        match variable {
            #[must_use] //~ WARN
            _match_arm => {}
        }

        let tail = ();

        #[must_use] //~ WARN
        tail
    }

    {
        fn f(_: impl Fn()) {}
        // Is this attribute argument-level or expression-level?
        f(
            #[must_use] //~ WARN
            || {},
        );
    }

    #[must_use] //~ WARN
    unsafe {
        let _x = [1, 2, 3].get_unchecked(1);
    }
}

#[must_use] //~ WARN
unsafe extern "C" {
    #![must_use] //~ WARN
    //~^ WARN unused attribute
    //~^^ WARN this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!

    #[must_use] // OK
    pub fn external_function(
        #[must_use] arg: *mut u8, //~ ERROR
        #[must_use] ... //~ ERROR
    );

    #[must_use] //~ WARN
    pub static EXTERNAL_STATIC: *const u32;

    #[must_use] //~ WARN
    pub static mut EXTERNAL_STATIC_MUT: *mut u32;
}

#[must_use] // OK
pub unsafe extern "C" fn abi_function(#[must_use] _: u32) {} //~ ERROR
//~^ WARN
#[must_use] //~ WARN
#[macro_export]
macro_rules! my_macro {
    () => {};
}

#[must_use] // OK
#[test]
fn test_fn() {}
