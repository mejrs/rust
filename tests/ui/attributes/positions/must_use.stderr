error[E0658]: attributes on expressions are experimental
  --> $DIR/must_use.rs:73:13
   |
LL |             #[must_use]
   |             ^^^^^^^^^^^
   |
   = note: see issue #15701 <https://github.com/rust-lang/rust/issues/15701> for more information
   = help: add `#![feature(stmt_expr_attributes)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0658]: attributes on expressions are experimental
  --> $DIR/must_use.rs:138:20
   |
LL |     let _closure = #[must_use]
   |                    ^^^^^^^^^^^
   |
   = note: see issue #15701 <https://github.com/rust-lang/rust/issues/15701> for more information
   = help: add `#![feature(stmt_expr_attributes)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0658]: attributes on expressions are experimental
  --> $DIR/must_use.rs:142:25
   |
LL |     let _move_closure = #[must_use]
   |                         ^^^^^^^^^^^
   |
   = note: see issue #15701 <https://github.com/rust-lang/rust/issues/15701> for more information
   = help: add `#![feature(stmt_expr_attributes)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0658]: attributes on expressions are experimental
  --> $DIR/must_use.rs:155:22
   |
LL |         let _array = #[must_use]
   |                      ^^^^^^^^^^^
   |
   = note: see issue #15701 <https://github.com/rust-lang/rust/issues/15701> for more information
   = help: add `#![feature(stmt_expr_attributes)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0658]: attributes on expressions are experimental
  --> $DIR/must_use.rs:162:22
   |
LL |         let _tuple = #[must_use]
   |                      ^^^^^^^^^^^
   |
   = note: see issue #15701 <https://github.com/rust-lang/rust/issues/15701> for more information
   = help: add `#![feature(stmt_expr_attributes)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error: allow, cfg, cfg_attr, deny, expect, forbid, and warn are the only allowed built-in attributes in function parameters
  --> $DIR/must_use.rs:71:23
   |
LL | ...   pub fn method(#[must_use] &mut self) {
   |                     ^^^^^^^^^^^

error: allow, cfg, cfg_attr, deny, expect, forbid, and warn are the only allowed built-in attributes in function parameters
  --> $DIR/must_use.rs:79:30
   |
LL | ...   pub fn static_method(#[must_use] _arg: u32) {}
   |                            ^^^^^^^^^^^

error: allow, cfg, cfg_attr, deny, expect, forbid, and warn are the only allowed built-in attributes in function parameters
  --> $DIR/must_use.rs:140:7
   |
LL |     | #[must_use] _arg: u32| {};
   |       ^^^^^^^^^^^

error: allow, cfg, cfg_attr, deny, expect, forbid, and warn are the only allowed built-in attributes in function parameters
  --> $DIR/must_use.rs:144:12
   |
LL |     move | #[must_use] _arg: u32| {};
   |            ^^^^^^^^^^^

error: allow, cfg, cfg_attr, deny, expect, forbid, and warn are the only allowed built-in attributes in function parameters
  --> $DIR/must_use.rs:224:9
   |
LL |         #[must_use] arg: *mut u8,
   |         ^^^^^^^^^^^

error: allow, cfg, cfg_attr, deny, expect, forbid, and warn are the only allowed built-in attributes in function parameters
  --> $DIR/must_use.rs:225:9
   |
LL |         #[must_use] ...
   |         ^^^^^^^^^^^

error: allow, cfg, cfg_attr, deny, expect, forbid, and warn are the only allowed built-in attributes in function parameters
  --> $DIR/must_use.rs:236:39
   |
LL | pub unsafe extern "C" fn abi_function(#[must_use] _: u32) {}
   |                                       ^^^^^^^^^^^

error[E0658]: trait aliases are experimental
  --> $DIR/must_use.rs:32:5
   |
LL |     trait TraitAlias = Debug;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #41517 <https://github.com/rust-lang/rust/issues/41517> for more information
   = help: add `#![feature(trait_alias)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

warning: `#[must_use]` has no effect when applied to an extern crate
  --> $DIR/must_use.rs:8:1
   |
LL | #[must_use]
   | ^^^^^^^^^^^
   |
note: the lint level is defined here
  --> $DIR/must_use.rs:4:9
   |
LL | #![warn(unused_attributes)]
   |         ^^^^^^^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a module
  --> $DIR/must_use.rs:13:1
   |
LL | #[must_use]
   | ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a module
  --> $DIR/must_use.rs:16:1
   |
LL | #[must_use]
   | ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a module
  --> $DIR/must_use.rs:119:5
   |
LL |     #![must_use]
   |     ^^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a closure
  --> $DIR/must_use.rs:138:20
   |
LL |     let _closure = #[must_use]
   |                    ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a function param
  --> $DIR/must_use.rs:140:7
   |
LL |     | #[must_use] _arg: u32| {};
   |       ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a closure
  --> $DIR/must_use.rs:142:25
   |
LL |     let _move_closure = #[must_use]
   |                         ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a function param
  --> $DIR/must_use.rs:144:12
   |
LL |     move | #[must_use] _arg: u32| {};
   |            ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an expression
  --> $DIR/must_use.rs:146:5
   |
LL |     #[must_use]
   |     ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an expression
  --> $DIR/must_use.rs:148:9
   |
LL |         #![must_use]
   |         ^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/must_use.rs:148:9
   |
LL |         #![must_use]
   |         ^^^^^^^^^^^^ help: remove this attribute
   |
note: attribute also specified here
  --> $DIR/must_use.rs:146:5
   |
LL |     #[must_use]
   |     ^^^^^^^^^^^
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!

warning: `#[must_use]` has no effect when applied to a statement
  --> $DIR/must_use.rs:152:9
   |
LL |         #[must_use]
   |         ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an expression
  --> $DIR/must_use.rs:155:22
   |
LL |         let _array = #[must_use]
   |                      ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an expression
  --> $DIR/must_use.rs:158:13
   |
LL |             #[must_use]
   |             ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an expression
  --> $DIR/must_use.rs:162:22
   |
LL |         let _tuple = #[must_use]
   |                      ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an expression
  --> $DIR/must_use.rs:165:13
   |
LL |             #[must_use]
   |             ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an expression
  --> $DIR/must_use.rs:170:13
   |
LL |             #[must_use]
   |             ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a struct field
  --> $DIR/must_use.rs:174:13
   |
LL |             #[must_use]
   |             ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a struct field
  --> $DIR/must_use.rs:180:13
   |
LL |             #[must_use]
   |             ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an expression
  --> $DIR/must_use.rs:185:13
   |
LL |             #[must_use]
   |             ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an expression
  --> $DIR/must_use.rs:189:9
   |
LL |         #[must_use]
   |         ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an match arm
  --> $DIR/must_use.rs:191:13
   |
LL |             #[must_use]
   |             ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an expression
  --> $DIR/must_use.rs:197:9
   |
LL |         #[must_use]
   |         ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a closure
  --> $DIR/must_use.rs:205:13
   |
LL |             #[must_use]
   |             ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an expression
  --> $DIR/must_use.rs:210:5
   |
LL |     #[must_use]
   |     ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a foreign module
  --> $DIR/must_use.rs:216:1
   |
LL | #[must_use]
   | ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a foreign module
  --> $DIR/must_use.rs:218:5
   |
LL |     #![must_use]
   |     ^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/must_use.rs:218:5
   |
LL |     #![must_use]
   |     ^^^^^^^^^^^^ help: remove this attribute
   |
note: attribute also specified here
  --> $DIR/must_use.rs:216:1
   |
LL | #[must_use]
   | ^^^^^^^^^^^
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!

warning: `#[must_use]` has no effect when applied to a function param
  --> $DIR/must_use.rs:236:39
   |
LL | pub unsafe extern "C" fn abi_function(#[must_use] _: u32) {}
   |                                       ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a macro def
  --> $DIR/must_use.rs:238:1
   |
LL | #[must_use]
   | ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a foreign static item
  --> $DIR/must_use.rs:228:5
   |
LL |     #[must_use]
   |     ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a foreign static item
  --> $DIR/must_use.rs:231:5
   |
LL |     #[must_use]
   |     ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a module
  --> $DIR/must_use.rs:6:1
   |
LL | #![must_use]
   | ^^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a static item
  --> $DIR/must_use.rs:19:5
   |
LL |     #[must_use]
   |     ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a static item
  --> $DIR/must_use.rs:22:5
   |
LL |     #[must_use]
   |     ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a constant item
  --> $DIR/must_use.rs:25:5
   |
LL |     #[must_use]
   |     ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a use
  --> $DIR/must_use.rs:28:5
   |
LL |     #[must_use]
   |     ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a trait alias
  --> $DIR/must_use.rs:31:5
   |
LL |     #[must_use]
   |     ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a type alias
  --> $DIR/must_use.rs:34:5
   |
LL |     #[must_use]
   |     ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a lifetime parameter
  --> $DIR/must_use.rs:39:9
   |
LL |         #[must_use] 'lt,
   |         ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a type parameter
  --> $DIR/must_use.rs:40:9
   |
LL |         #[must_use] T>
   |         ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a struct field
  --> $DIR/must_use.rs:42:9
   |
LL |         #[must_use]
   |         ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a struct field
  --> $DIR/must_use.rs:44:9
   |
LL |         #[must_use]
   |         ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a struct field
  --> $DIR/must_use.rs:49:28
   |
LL |     pub struct TupleStruct(#[must_use] pub u32);
   |                            ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a enum variant
  --> $DIR/must_use.rs:53:9
   |
LL |         #[must_use]
   |         ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a struct field
  --> $DIR/must_use.rs:55:13
   |
LL |             #[must_use]
   |             ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a enum variant
  --> $DIR/must_use.rs:58:9
   |
LL |         #[must_use]
   |         ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a struct field
  --> $DIR/must_use.rs:59:22
   |
LL |         TupleVariant(#[must_use] u32),
   |                      ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a struct field
  --> $DIR/must_use.rs:63:9
   |
LL |         #[must_use]
   |         ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an implementation block
  --> $DIR/must_use.rs:67:5
   |
LL |     #[must_use]
   |     ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an implementation block
  --> $DIR/must_use.rs:83:5
   |
LL |     #[must_use]
   |     ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a lifetime parameter
  --> $DIR/must_use.rs:85:9
   |
LL |         #[must_use] 'lt,
   |         ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a type parameter
  --> $DIR/must_use.rs:86:9
   |
LL |         #[must_use] T
   |         ^^^^^^^^^^^

warning: unused attribute
  --> $DIR/must_use.rs:98:9
   |
LL |         #![must_use]
   |         ^^^^^^^^^^^^ help: remove this attribute
   |
note: attribute also specified here
  --> $DIR/must_use.rs:94:5
   |
LL |     #[must_use] // OK
   |     ^^^^^^^^^^^
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!

warning: `#[must_use]` has no effect when applied to a const parameter
  --> $DIR/must_use.rs:96:9
   |
LL |         #[must_use] const N: usize
   |         ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an expression
  --> $DIR/must_use.rs:111:9
   |
LL |         #[must_use]
   |         ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a function param
  --> $DIR/must_use.rs:71:23
   |
LL | ...   pub fn method(#[must_use] &mut self) {
   |                     ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an expression
  --> $DIR/must_use.rs:73:13
   |
LL |             #[must_use]
   |             ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a function param
  --> $DIR/must_use.rs:79:30
   |
LL | ...   pub fn static_method(#[must_use] _arg: u32) {}
   |                            ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to a provided trait method
  --> $DIR/must_use.rs:88:9
   |
LL |         #[must_use]
   |         ^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an implementation block
  --> $DIR/must_use.rs:122:9
   |
LL |         #![must_use]
   |         ^^^^^^^^^^^^

warning: `#[must_use]` has no effect when applied to an expression
  --> $DIR/must_use.rs:131:13
   |
LL |             #![must_use]
   |             ^^^^^^^^^^^^

error: aborting due to 13 previous errors; 62 warnings emitted

For more information about this error, try `rustc --explain E0658`.
